### ISSUE TITLE
[proposal] expose internal date pattern resolved per instance

### ISSUE THREAD

**Issue by caridy** [tc39/ecma402#21]
As part of the effort to align with EWM, we want to expose more internals, low level APIs. The resolved pattern for dates is something V8 is already exposing, where it matches the CLDR pattern used to format the date. E.g.:

```
(new Intl.DateTimeFormat("en-US")).resolved.pattern // should return "M/d/y"
```

From the user perspective, having access to the CLDR pattern helps to interoperate with other systems and languages, including the DOM. E.g.: specifically the `<input/>` tag where the pattern can be specified.

**Comment by caridy** [comment]
backpointer to an old task related to this proposal: https://bugs.ecmascript.org/show_bug.cgi?id=771

**Comment by zbraniecki** [comment]
Example of what this API would save us from doing: https://github.com/mozilla-b2g/gaia/blob/1360b05876d2def34a37c28e44d49e7e60dc2484/apps/system/js/value_selector/spin_date_picker.js#L120-L156

**Comment by caridy** [comment]
I need to talk to @bterlson about this to see how realistic this will be for Chakra/Edge since they don't use CLDR data.

**Comment by zbraniecki** [comment]
Is DateTimeFormat using CLDR? I'm not sure. I'm a bit confused about how CLDR handles order of "hma" (where h is hour, m i minute and a is dayperiod) because there doesn't seem to be an entry with that name. So it seems that CLDR for locales that don't use dayperiod by default, doesn't provide a way to order *with* dayperiod. `DateTimeFormat` handles that, so it may be coming from somewhere else.

Also, Mozilla would be very interested in such data for our date/time pickers.

**Comment by caridy** [comment]
@zbraniecki for that particular case, they use `hm` instead of `hma`, which can be matched with the fact that for `en` the value of `hour12` config is true by default. In other words, yes, it is the raw CLDR data:

```
"hm": "h:mm a"
```

while for `en-GB` we have:

```
"Hm": "HH:mm",
```

matching the default value for `hour12`, which if false.

**Comment by caridy** [comment]
I don't think I have answered your question, let me try again. the algo we have today does NOT use the key, it uses the value from CLDR to find the best match based on the provided configuration, which means we will either pickup entry `hm` or `Hm` depending of the value of `hour12` (whether that's provided in the instance, or the default value per locale), in other words:

```javascript
(new Intl.DateTimeFormat("en-US", { hour: true, minute: true }));
```

will use `h:mm a` for `en`, and `HH:mm` for `en-GB`, if you add `hour12` to that config, you can choose which one to use by forcing the value of it.

**Comment by zbraniecki** [comment]
I'm going to expose it internally for Gecko because we need this for date/time pickers.

I *may* be interested in championing it, but I'm a bit afraid of the scope of spec work to define all combinations that `pattern` can take depending on locale and options.

@caridy - am I correct to expect it to be a substantial amount of work or would you be ok to leave it as implementation specific?

**Comment by zbraniecki** [comment]
Alternatively, one can use `formatToParts` to build a placeholder:

```javascript
var x = new Intl.DateTimeFormat('en-US', {
  hour: '2-digit',
  minute: '2-digit',
});
var parts = x.formatToParts(now);

parts.map(part => {
  switch (part.type) {
    case 'hour':
    case 'minute':
      return '--'
    default:
      return part.value;
  }
}).join('');
```

**Comment by zbraniecki** [comment]
I guess my question is - except of building placeholder patterns (which can be done using `formatToParts`), is there any other use case for the pattern?

**Comment by zbraniecki** [comment]
@eemli, @rxaviers, @ericf, @littledan - any opinion on that?

If the only use case for the pattern is to create placeholders or pickers and we can already do that with `formatToParts`, I suggest we close this issue.

**Comment by eemeli** [comment]
I at least can't think of anything that could be done better with `resolved.pattern` rather than `formatToParts`.

**Comment by zbraniecki** [comment]
then I suggest we close this issue! :) @caridy, can you do this? I don't have the privs.

**Comment by caridy** [comment]
Yes, in fact chrome does not longer expose the internals of the DateTimeFormat, there is no reason to do so now that we have `formatToParts`.

**Comment by rxaviers** [comment]
:+1: (although late :blush:)
